File: code/qa_motion.R
library(tidyverse)
library(dplyr)
# Check motion regressors

exclude_subs=c('sub-204')

# path to confounds folder
fmriprep_dir <- '/Users/miriam/Documents/hoffman2/bids/fmriprep'
setwd(fmriprep_dir)
# list all directories in this folder (subject directories)
all_dirs <- list.dirs(path = ".", full.names = FALSE, recursive = FALSE)
# combine all subjects' confounds tsv files
all_dfs <- c()
for (d in all_dirs) {
  # if this is a subjects folder...
  if (startsWith(d,'sub-')){
    # subject ID
    sub <- d
    if (sub %in% exclude_subs){
      print(paste('Excluding subject ', sub))
    } else {
      # find all tsv files for this subject
      confound_fnames <- list.files(path = paste0('./',d,'/func'),pattern='*tsv', full.names = TRUE)
      # setup this subject's dataframe
      sub_df=c()
      for (f in confound_fnames){
        print(paste('Reading in file:',f))
        # index for task name in filename
        task_i <- regexpr("task-", f)[1]+5
        # task name
        task <- substring(f,task_i,task_i+5)
        # index for run number in filename
        run_i <- regexpr("run-", f)[1]+4
        # run number
        run <- substr(f,run_i,run_i+1)
        # read in this file
        c <- read.csv(f, sep = '\t', na.strings = 'n/a')
        # subselect movement parameters
        c <- select(c, dvars, framewise_displacement, trans_x, trans_y, trans_z,
                        rot_x, rot_y, rot_z)
        # add task name to dataframe
        c$task <- task
        # add run number to dataframe
        c$run <- run
        # combine this file's data to subject's dataframe
        sub_df <- rbind(sub_df, c)
      }
      # add subject's ID to dataframe
      sub_df$subID <- sub
      # append subject's dataframe to all_dfs
      all_dfs <- rbind(all_dfs, sub_df)
    }
  }
}

# get all subjects
subs <- unique(all_dfs$subID)
# get all tasks
tasks <- unique(all_dfs$task)
# get all runs
runs <- unique(all_dfs$run)

# create new dataframe with one row per run per subject
summ_df <- expand.grid(subs, runs)
colnames(summ_df) <- c('subID', 'run')
# go through each run for each subject, calculate change between frames, 
# determine if these movements are too big (>2 or >.5), and count how many
# of each subject's runs fail this test
all_summ <- data.frame()
for (r in seq(nrow(summ_df))){
  # subject ID
  sub_curr <- summ_df[r,'subID']
  # run number
  run_curr <- summ_df[r,'run']
  # data from this run for this subject
  d <- filter(all_dfs, subID==sub_curr, run==run_curr)
  # task name
  task_curr <- unique(d$task)
  # calculate frame-by-frame changes
  # save the distance from the first frame
  first_diff <- sqrt(d[1,'trans_x']**2 + d[1,'trans_y']**2 + d[1,'trans_z']**2)
  # calculate differences between each frame and the previous frame
  xyz_diff <- data.frame(diff(as.matrix(select(d, trans_x, trans_y, trans_z))))
  # calculate what this difference is in terms of distance
  d$trans_dist <- c(first_diff, sqrt(xyz_diff$trans_x**2 + xyz_diff$trans_y**2 + 
                             xyz_diff$trans_z**2))
  # is it bigger than 2mm?
  d$gt2mm <- as.numeric(d$trans_dist > 2)
  # is it bigger than .5mm?
  d$gt_5mm <- as.numeric(d$trans_dist > .5)
  # does this run fail?
  summ <- ddply(d, .(subID, task, run), summarise, 
                n2=sum(gt2mm), n_5=sum(gt_5mm),
                fail=as.numeric(n2>=1 || n_5 >= 5))
  # add to total summary
  all_summ <- rbind(all_summ, summ)
}

write_csv(all_summ, paste0(fmriprep_dir,'/motion_counts.csv'))
# count how many runs fail per subject per task
out <- ddply(all_summ, .(subID, task), summarize, fails=sum(fail))
write_csv(out, paste0(fmriprep_dir,'/motion_summary.csv'))
# if at least half of the runs (i.e. 2) fail, add to failed subjects list
failed_subs <- filter(out, fails>=2)$subID
# print out failed subjects
print(paste(length(failed_subs), 'failed subjects:', failed_subs))

      

File: code/qa_motion.py
#!/bin/python

import os
import shutil
import sys
from datetime import datetime
from copy import deepcopy
import glob
import pandas as pd
import nibabel as nib
import numpy as np
from scipy.stats import ttest_rel, ttest_ind, ttest_1samp
from funcs import *

def get_term(fname, key):
    fname = os.path.basename(fname).split('.')[0]
    components = fname.split('_')
    key = key.split('-')[0] + '-'
    for i, s in enumerate(components):
        if key in s:
            return s.split('-')[1]
    return -1

print(str(datetime.now()) + ": Begin qa_motion.py")

# read in arguments
# subjects
subs = sys.argv[1:]
if len(subs) == 0:
    subs='all'
subs_str = ""

in_dir = FMRIPREP_DIR
out_dir = FMRIPREP_DIR

# column names for translational motion regressors
motion_columns = ["trans_x", "trans_y", "trans_z"]

# list all directories in this folder (subject directories)
dnames = os.path.join(in_dir,
                         "sub-???")
all_fmriprep_dirs = glob.glob(dnames)
print(all_fmriprep_dirs)
# list of subjects who fail motion QA
fail_subs = []
# combine all subjects' confounds tsv files
for d, dir in enumerate(all_fmriprep_dirs):
    sub = os.path.basename(dir)
    print("Analyzing " + sub)
    if sub in subs or subs == "all":
        if sub in EXCLUDE_SUBS:
            print('Excluding subject '+ sub)
        else:
            subs_str += sub + "_"
            sub_dir = os.path.join(in_dir,sub,'func')
            # find all tsv files for this subject
            confound_fnames = glob.glob(os.path.join(sub_dir, '*desc-confounds_regressors.tsv'))
            # count number of runs this subject fails
            sub_fails = 0
            # construct this subject's dataframe
            for i, f in enumerate(confound_fnames):
                print('Reading in file ' + f)
                # get task
                task = get_term(f, 'task')
                # get run
                run = get_term(f, 'run')
                # read in file
                raw_df = pd.read_csv(f, delimiter = '\t')
                # select specific columns
                sel_df = raw_df[motion_columns]
                # get number of rows
                nrow = sel_df.shape[0]

                # setup columns for distances
                sel_df['dist'] = None
                sel_df['mm2.0'] = 0
                sel_df['mm0.5'] = 0
                # calculate movement
                print("Calculating distances")
                diff = np.zeros((1,nrow))
                for m in motion_columns:
                    col = sel_df[m]
                    d = [0]+[j-i for i, j in zip(col[:-1], col[1:])]
                    d = np.square(d)
                    diff += d
                dist = np.sqrt(diff)
                dist = np.transpose(dist)
                sel_df['dist'] = dist
                sel_df['mm2.0'] = dist > 2
                sel_df['mm0.5'] = dist > .5

                sum20 = sum(sel_df['mm2.0'])
                sum05 = sum(sel_df['mm0.5'])
                fail = sum20>0 or sum05 >= 5
                if fail:
                    print("Subject "+sub+" fails on run "+run)
                sub_fails += int(fail)
                run_dict = {'sub':[sub], 'task':[task], 'run':[run],
                          'dist':[np.mean(sel_df['dist'])], 'mm2.0':[sum20],
                          'mm0.5':[sum05], 'run_fail':[fail]}
                if i == 0:
                    sub_df = pd.DataFrame(run_dict)
                else:
                    sub_df = sub_df.append(pd.DataFrame(run_dict))

        print("Subject " + sub + " failed " + str(sub_fails) + " runs.")
        if sub_fails > 4:
            fail_subs += [sub]
            sub_df['sub_fail'] = True
        else:
            sub_df['sub_fail'] = False
        # add subject's dataframe to the full dataframe
        if 'full_df' in locals():
            full_df = full_df.append(sub_df)
        else:
            full_df = deepcopy(sub_df)
# save output
fname = os.path.join(out_dir, subs_str + "motion_counts.csv")
full_df.to_csv(fname)
print("Saved " + fname)
